src/hotspot/share/runtime/thread.cpp

```diff
@@ -3900,3 +3900,40 @@ void JavaThread::verify_cross_modify_fence_failure(JavaThread *thread) {
    report_vm_error(__FILE__, __LINE__, "Cross modify fence failure", "%p", thread);
 }
 #endif
+
+// Starts the target JavaThread as a daemon of the given priority, and
+// bound to the given java.lang.Thread instance.
+// The Threads_lock is held for the duration.
+void JavaThread::start_internal_daemon(JavaThread* current, JavaThread* target,
+                                       Handle thread_oop, ThreadPriority prio) {
+  MutexLocker mu(current, Threads_lock);
+
+  // Initialize the fields of the thread_oop first
+
+  java_lang_Thread::set_thread(thread_oop(), target); // isAlive == true now
+
+  if (prio != NoPriority) {
+    java_lang_Thread::set_priority(thread_oop(), prio);
+    // Note: we don't call os::set_priority here. Possibly we should,
+    // else all threads should call it themselves when they first run.
+  }
+
+  java_lang_Thread::set_daemon(thread_oop());
+
+  // Now bind the thread_oop to the target JavaThread
+  target->set_threadObj(thread_oop());
+
+  Threads::add(target); // target is now visible for safepoint/handshake
+  Thread::start(target);
+}
+
+void JavaThread::vm_exit_on_thread_allocation_failure(JavaThread* thread) {
+  // At this point it may be possible that no osthread was created for the
+  // JavaThread due to lack of memory. We would have to throw an exception
+  // in that case. However, since this must work and we do not allow
+  // exceptions anyway, check and abort if this fails.
+  if (thread == nullptr || thread->osthread() == nullptr) {
```

Comment:
thread shouldn't be NULL here if you haven't used a nothrow version of new to allocate the thread.

Reply:
Since you're no longer holding the JavaThreads_lock here, it seems like you could do this from the JavaThread constructor instead of adding this function. Except for the case for the JFR thread and JVMTI. So maybe you need this.  Maybe fix the comment at the end of JavaThread constructor that says you're holding a lock and are going to throw OOM and those details.